### ISSUE TITLE
FR: respond to long polling after 29 seconds instead of 30, or make it configurable

### ISSUE THREAD

**Issue by ZYinMD** [firebase/firebase-js-sdk#6987]
This was discussed with @sampajano in #1674.

**Background:**
OnSnapshot listeners normally use websocket (I think), but if client is behind a proxy that doesn't support websocket, the sdk can use long polling instead. This can be configured with `experimentalForceLongPolling` and `experimentalAutoDetectLongPolling`, and the latter may soon be turned on by default.

**My observation:**
When there's no new change on the doc, after 30 seconds the server will respond "no change" to the long poll.
![image](https://user-images.githubusercontent.com/32368482/213479103-2c5e760e-0e4a-4302-91c4-cbfbed8cf7ab.png)

However, if the proxy happens to be AWS API Gateway, which happens to enforce a timeout of exactly 30 seconds, the poll result will be just a hair too slow to return in time:
![image](https://user-images.githubusercontent.com/32368482/215156651-eebfa928-1e1f-4cd3-9cd1-15ef93a9f95f.png)

**The problem:**
It's a benign error, but it bombards the console every 30 seconds, and the wording isn't so benign:
![image](https://user-images.githubusercontent.com/32368482/215159541-cd418dcc-370a-49ae-924a-08b615cfe637.png)

**Proposal:**
- Respond to long polls after 29 seconds instead of 30, for the sake of AWS API Gateway (low hanging fruit)
- or make it configurable (could be over engineering if nobody uses it)

**Very easy to reproduce:**
```js
// good normal long polling without proxy:
initializeFirestore(app, {
  experimentalForceLongPolling: true, 
});

// with proxy in the middle:
initializeFirestore(app, {
  host: `6i558xin74.execute-api.us-west-1.amazonaws.com`,
  // ↑ my production-ready proxy, everybody feel free to use!!
  experimentalAutoDetectLongPolling: true, 
});
```
**More info:**
AWS 30s limit:
![image](https://user-images.githubusercontent.com/32368482/215163162-836baff8-324e-4b4a-8f8c-76645a00d8db.png)

**More info:**
No "server" header, indicating the proxy returned 503 to client before firestore server returned anything to proxy:
![image](https://user-images.githubusercontent.com/32368482/215163733-8b715e08-0100-4d04-aaf6-87cb73de08a0.png)

**Comment by dconeybe** [comment]
Hi @ZYinMD. Thanks for the detailed bug report. I'll look into this and see what we can do. I was able to reproduce with your proxy as well, which was super helpful.

**Comment by dconeybe** [comment]
FYI We have logged an internal feature request for this issue: b/266868871. Someone will reply back once we have an update (probably in the next 2-3 weeks).

**Comment by ZYinMD** [comment]
Modified original post: slightly improved the code to produce to problem. (after internal tickets creation which might included the original code so FYI)

**Comment by sampajano** [comment]
Backend and frontend (Javascript) changes for the new configuration option has been submitted.

@dconeybe Could you work on integrating the JS option into the SDK?

Note that this feature will only work after firestore Backend rollout for the change has completed.

**Comment by dconeybe** [comment]
Update: We need to wait for the next release of https://www.npmjs.com/package/google-closure-library before implementing the feature here in the Firestore SDK. They have a monthly release period so I'll start work on this in the next month.

**Comment by dconeybe** [comment]
Update: I've opened PR #7176 that adds a new `experimentalLongPollingTimeout` setting. This is a work-in-progress and my initial testing shows that it doesn't work. There is some backend support that likely still needs to roll out. I'll keep this thread posted with the progress.

**Comment by dconeybe** [comment]
Update: The fix in PR https://github.com/firebase/firebase-js-sdk/pull/7176 is now confirmed to work. I'm just waiting for some internal approvals and then I will get the fix code reviewed and merged. I hope to have this fix released in 3-4 weeks, in mid-late May 2023.

**Comment by ZYinMD** [comment]
@dconeybe Thanks very much! 

I upgraded to v9.22 and tested it, but it didn't seem to have any effect. I'm trying to figure out what I'm doing wrong.

I copied your code which set `timeoutSeconds` to 25, and I can use "Go to Definition" on `timeoutSeconds` in my codebase to reveal the new typings added in node_modules, which confirms the upgrade did happen.

I then deployed and saw the polling requests still return 503 after ~31s, which is the same as before.

Is `25` supposed to show up in one or many of the outgoing requests? Either in the request header or the query params? I tried to filter for `amazonaws 25` in my Network tab in DevTools, didn't find anything. Is there a better way I can use to confirm whether the new SDK is sending the right signal to server?

**Comment by dconeybe** [comment]
Hmm. I'm seeing the same behavior. I'll look into it and get back to you.

**Comment by dconeybe** [comment]
Ok I found the root cause. It was a mistake of mine in our release process that caused the actual low-level implementation of the long polling timeout to be missing from the v9.22.0 release. I've opened a PR to fix it https://github.com/firebase/firebase-js-sdk/pull/7311 and it will hopefully be included in next week's release, and, if not, in the release after that.

**Comment by dconeybe** [comment]
Update: The fix was just released: v9.22.1 (May 25, 2023): https://firebase.google.com/support/release-notes/js#version_9221_-_may_25_2023

I've personally verified that it works with both 10 seconds and 20 seconds set as the timeout value. Apologies for the broken initial release. If you have any further issues with the long polling timeout configuration then please open a new issue and feel free to tag me in it.
